File: cli/src/commands.ts
import { createInterface } from 'node:readline/promises'
import { cp, mkdir, readFile, rm, stat, writeFile } from 'node:fs/promises'
import { basename, dirname, join } from 'node:path'
import { realpath } from 'node:fs/promises'
import { ConfigFile, parseYaml, stringifyYaml } from './config-io.js'
import { ipcCall } from './ipc.js'
import { hr, humanBytes, pad, row, statusBadge, timeAgo } from './format.js'
import type {
  FileListEntry,
  FileVersionEntry,
  LogEntry,
  PluginInfo,
  TaskDetail,
  TaskSummary,
  VersionInfo,
} from '@edge-sync/protocol'

export interface Ctx {
  cfg: ConfigFile
}

const W = [16, 10, 12, 12, 14] as const

export async function cmdList(ctx: Ctx): Promise<void> {
  const tasks = await ipcCall<TaskSummary[]>(ctx.cfg.socketPath(), 'task.list')
  console.log(row([
    ['NAME', W[0]], ['STATUS', W[1]], ['PLUGIN', W[2]], ['INTERVAL', W[3]], ['LAST SYNC', W[4]],
  ]))
  console.log(hr(70))
  for (const t of tasks) {
    const failed = t.consecutiveFailures > 0 ? ` (${t.consecutiveFailures} fails)` : ''
    console.log(row([
      [t.name, W[0]],
      [statusBadge(t.status + failed), W[1] + 14],
      [t.plugin, W[2]],
      [t.interval, W[3]],
      [timeAgo(t.lastSuccessAt), W[4]],
    ]))
  }
  console.log(`\n${tasks.length} task(s)`)
}

export async function cmdStatus(ctx: Ctx, name?: string): Promise<void> {
  if (!name) {
    await cmdList(ctx)
    return
  }
  const d = await ipcCall<TaskDetail>(ctx.cfg.socketPath(), 'task.status', { name })
  console.log(`task:      ${d.name}`)
  console.log(`plugin:    ${d.plugin}`)
  console.log(`status:    ${statusBadge(d.status)}`)
  console.log(`interval:  ${d.interval}`)
  console.log(`next run:  ${d.nextRunAt ?? '-'}`)
  console.log(`last sync: ${d.lastSyncAt ?? 'never'} (success: ${d.lastSuccessAt ?? 'never'})`)
  console.log(`failures:  ${d.consecutiveFailures}`)
  console.log(`retention: keepLast=${d.retention.keepLast} keepDays=${d.retention.keepDays}`)
  console.log(`stats:     ${d.stats.totalSyncs} syncs, ${d.stats.totalFiles} files, ${humanBytes(d.stats.totalBytes)}`)
}

export async function cmdSync(ctx: Ctx, name?: string): Promise<void> {
  if (!name) {
    console.error('usage: edge-sync sync <task>')
    process.exitCode = 1
    return
  }
  console.log(`syncing ${name} ...`)
  const res = await ipcCall<{ triggered: boolean }>(ctx.cfg.socketPath(), 'task.trigger', { name })
  if (res.triggered) {
    console.log(`${name}: sync completed`)
  }
}

export async function cmdHistory(ctx: Ctx, name?: string): Promise<void> {
  if (!name) {
    console.error('usage: edge-sync history <task>')
    process.exitCode = 1
    return
  }
  const versions = await ipcCall<VersionInfo[]>(ctx.cfg.socketPath(), 'history.list', { name })
  console.log(row([
    ['VERSION', 20], ['FILES', 8], ['SIZE', 12], ['NOTE', 10],
  ]))
  console.log(hr(56))
  for (const v of versions) {
    console.log(row([
      [v.version, 20],
      [String(v.files), 8],
      [humanBytes(v.bytes), 12],
      [v.isCurrent ? '<- current' : '', 10],
    ]))
  }
  console.log(`\n${versions.length} version(s)`)
}

export async function cmdLogs(ctx: Ctx, n = 30, task?: string): Promise<void> {
  const entries = await ipcCall<LogEntry[]>(ctx.cfg.socketPath(), 'log.tail', { n, task: task ?? '' })
  for (const e of entries) {
    const attrs = e.attrs ? `  ${e.attrs}` : ''
    console.log(`${e.time} ${pad(e.level.toUpperCase(), 6)} ${e.msg}${attrs}`)
  }
}

export async function cmdPlugins(ctx: Ctx): Promise<void> {
  const plugins = await ipcCall<PluginInfo[]>(ctx.cfg.socketPath(), 'plugin.list')
  for (const p of plugins) {
    console.log(`${pad(p.name, 12)} v${p.version}`)
    if (p.configSchema) {
      const schema = p.configSchema as { properties?: Record<string, { type?: string; description?: string }> }
      for (const [k, v] of Object.entries(schema.properties ?? {})) {
        console.log(`  - ${k} (${v.type ?? 'any'})${v.description ? `: ${v.description}` : ''}`)
      }
    }
  }
}

// ---------- add 向导 ----------

async function ask(rl: ReturnType<typeof createInterface>, question: string, def?: string): Promise<string> {
  const suffix = def !== undefined && def !== '' ? ` [${def}]` : ''
  const a = (await rl.question(`${question}${suffix}: `)).trim()
  return a === '' ? (def ?? '') : a
}

export async function cmdAdd(ctx: Ctx): Promise<void> {
  const rl = createInterface({ input: process.stdin, output: process.stdout })
  try {
    const plugins = await ipcCall<PluginInfo[]>(ctx.cfg.socketPath(), 'plugin.list')
    if (plugins.length === 0) {
      console.error('no plugins discovered in pluginBinDir')
      return
    }
    console.log('available plugins:')
    plugins.forEach((p, i) => console.log(`  ${i + 1}. ${p.name} (v${p.version})`))

    let plugin = ''
    for (;;) {
      const pick = await ask(rl, 'plugin (number or name)')
      const idx = Number.parseInt(pick, 10)
      const chosen = Number.isInteger(idx) && idx >= 1 && idx <= plugins.length
        ? plugins[idx - 1]
        : plugins.find((p) => p.name === pick)
      if (chosen) {
        plugin = chosen.name
        break
      }
      console.log('  invalid choice')
    }

    const chosen = plugins.find((p) => p.name === plugin)!
    const name = await ask(rl, 'task name (letters/digits/._-)')
    if (!/^[a-zA-Z0-9][a-zA-Z0-9._-]*$/.test(name)) {
      console.error('invalid task name')
      process.exitCode = 1
      return
    }
    const intervalSec = Number.parseInt(await ask(rl, 'poll interval (seconds)', '600'), 10)
    const keepLast = Number.parseInt(await ask(rl, 'retention keepLast (versions to keep)', '10'), 10)
    const keepDays = Number.parseInt(await ask(rl, 'retention keepDays (0 = unlimited)', '0'), 10)

    const options: Record<string, unknown> = {}
    const props = (chosen.configSchema as { properties?: Record<string, { type?: string; description?: string }> })
      ?.properties
    if (props && Object.keys(props).length > 0) {
      console.log(`plugin ${plugin} options:`)
      for (const [key, prop] of Object.entries(props)) {
        const desc = prop.description ? ` (${prop.description})` : ''
        const raw = await ask(rl, `  ${key} [${prop.type ?? 'any'}]${desc}`)
        if (raw === '') continue
        options[key] = prop.type === 'number' || prop.type === 'integer'
          ? Number(raw)
          : prop.type === 'boolean'
            ? raw === 'true' || raw === 'yes'
            : raw
      }
    } else {
      const raw = await ask(rl, 'options (JSON, empty = {})', '{}')
      Object.assign(options, parseYaml(raw) as Record<string, unknown>)
    }

    const task = {
      name,
      plugin,
      interval: `${intervalSec}s`,
      options,
      retention: { keepLast, keepDays },
    }
    console.log('\n' + stringifyYaml({ tasks: [task] }))
    const confirm = await ask(rl, 'write this task to config? (y/N)', 'n')
    if (confirm.toLowerCase() !== 'y') {
      console.log('aborted')
      return
    }
    const tasks = ctx.cfg.tasks().filter((t) => t['name'] !== name)
    tasks.push(task)
    ctx.cfg.data.tasks = tasks
    ctx.cfg.save()
    await reload(ctx)
    console.log(`task "${name}" added`)
  } finally {
    rl.close()
  }
}

export async function cmdEdit(ctx: Ctx, name?: string): Promise<void> {
  if (!name) {
    console.error('usage: edge-sync edit <task>')
    process.exitCode = 1
    return
  }
  const tasks = ctx.cfg.tasks()
  const idx = tasks.findIndex((t) => t['name'] === name)
  if (idx < 0) {
    console.error(`task not found: ${name}`)
    process.exitCode = 1
    return
  }
  const editor = process.env['EDITOR'] || 'vi'
  const tmp = `${ctx.cfg.path}.edit-${name}.yml`
  await writeFile(tmp, `# edit task "${name}"; name must stay unchanged\n${stringifyYaml(tasks[idx])}`)
  const { spawnSync } = await import('node:child_process')
  const res = spawnSync(editor, [tmp], { stdio: 'inherit' })
  if (res.status !== 0) {
    console.error('editor exited nonzero, aborting')
    return
  }
  let edited: Record<string, unknown>
  try {
    edited = parseYaml(await readFile(tmp, 'utf8')) as Record<string, unknown>
  } catch (e) {
    console.error(`parse edited file failed: ${(e as Error).message}`)
    await rm(tmp, { force: true })
    process.exitCode = 1
    return
  }
  await rm(tmp, { force: true })
  if (edited['name'] !== name) {
    console.error('task name changed in editor, aborting')
    process.exitCode = 1
    return
  }
  tasks[idx] = edited
  ctx.cfg.data.tasks = tasks
  ctx.cfg.save()
  await reload(ctx)
  console.log(`task "${name}" updated`)
}

export async function cmdRemove(ctx: Ctx, name?: string, flags: Map<string, string> = new Map()): Promise<void> {
  if (!name) {
    console.error('usage: edge-sync remove <task> [--purge]')
    process.exitCode = 1
    return
  }
  const tasks = ctx.cfg.tasks()
  const remaining = tasks.filter((t) => t['name'] !== name)
  if (remaining.length === tasks.length) {
    console.error(`task not found: ${name}`)
    process.exitCode = 1
    return
  }
  if (flags.get('purge') === 'true') {
    const dir = join(ctx.cfg.storagePath('dataDir'), name)
    console.log(`purging data directory: ${dir}`)
    await rm(dir, { recursive: true, force: true })
  }
  ctx.cfg.data.tasks = remaining
  ctx.cfg.save()
  await reload(ctx)
  console.log(`task "${name}" removed`)
}

export async function cmdExport(ctx: Ctx, name?: string, flags: Map<string, string> = new Map()): Promise<void> {
  if (!name || !flags.get('out')) {
    console.error('usage: edge-sync export <task> [--version <ts>] --out <dir>')
    process.exitCode = 1
    return
  }
  const dataDir = join(ctx.cfg.storagePath('dataDir'), name)
  const version = flags.get('version') || 'current'
  const src = version === 'current'
    ? join(dataDir, 'current')
    : join(dataDir, 'versions', version)

  await stat(src) // 不存在直接抛错
  const out = flags.get('out')!
  await mkdir(out, { recursive: true })
  // current 是 symlink，解引用为真实版本目录再拷贝。
  const realSrc = await realpath(src)
  await cp(realSrc, out, {
    recursive: true,
    filter: (s) => basename(s) !== '.edge-sync-manifest.json',
  })
  console.log(`exported ${name}@${version} -> ${out}`)
}

export async function cmdRestoreFile(
  ctx: Ctx,
  name?: string,
  path?: string,
  flags: Map<string, string> = new Map(),
): Promise<void> {
  if (!name || !path || !flags.get('out')) {
    console.error('usage: edge-sync restore-file <task> <path> [--version <ts>] --out <file>')
    process.exitCode = 1
    return
  }
  const dataDir = join(ctx.cfg.storagePath('dataDir'), name)
  const version = flags.get('version') || 'current'
  const src = version === 'current'
    ? join(dataDir, 'current', path)
    : join(dataDir, 'versions', version, path)
  const content = await readFile(src)
  const out = flags.get('out')!
  await mkdir(dirname(out), { recursive: true })
  await writeFile(out, content)
  console.log(`restored ${name}@${version}:${path} -> ${out} (${content.length} bytes)`)
}

export async function cmdFileVersions(ctx: Ctx, name?: string, path?: string): Promise<void> {
  if (!name || !path) {
    console.error('usage: edge-sync file-versions <task> <path>')
    process.exitCode = 1
    return
  }
  const versions = await ipcCall<FileVersionEntry[]>(
    ctx.cfg.socketPath(), 'history.fileVersions', { name, path })
  console.log(row([['VERSION', 20], ['SIZE', 12], ['FINGERPRINT', 16]]))
  console.log(hr(52))
  for (const v of versions) {
    console.log(row([
      [v.version, 20],
      [humanBytes(v.size ?? 0), 12],
      [v.fingerprint.slice(0, 14), 16],
    ]))
  }
}

export async function cmdReload(ctx: Ctx): Promise<void> {
  await reload(ctx)
}

async function reload(ctx: Ctx): Promise<void> {
  const res = await ipcCall<{ reloaded: boolean; tasks: number }>(
    ctx.cfg.socketPath(), 'config.reload')
  console.log(`config reloaded (${res.tasks} task(s))`)
}

File: cmd/edge-sync/commands.go
// 全部 CLI 命令实现（对齐 TS 版 commands.ts）。
package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"

	"edge-sync/pkg/protocol"
)

type ctx = *context

func needTask(name, usage string) (string, error) {
	if name == "" {
		return "", errors.New(usage)
	}
	return name, nil
}

func cmdList(ctx ctx) error {
	var tasks []protocol.TaskSummary
	if err := ctx.ipc().call("task.list", map[string]any{}, &tasks); err != nil {
		return err
	}
	fmt.Println(pad("NAME", 16), pad("STATUS", 24), pad("PLUGIN", 10), pad("INTERVAL", 10), "LAST SYNC")
	fmt.Println(hr(72))
	for _, t := range tasks {
		status := t.Status
		if !t.Enabled {
			status = "paused"
		} else if t.ConsecutiveFailures > 0 {
			status = fmt.Sprintf("%s (%d fails)", status, t.ConsecutiveFailures)
		}
		fmt.Println(pad(t.Name, 16), pad(status, 24), pad(t.Plugin, 10), pad(t.Interval, 10), timeAgo(t.LastSuccessAt))
	}
	fmt.Printf("\n%d task(s)\n", len(tasks))
	return nil
}

func cmdStatus(ctx ctx, name string) error {
	if name == "" {
		return cmdList(ctx)
	}
	var d protocol.TaskDetail
	if err := ctx.ipc().call("task.status", map[string]any{"name": name}, &d); err != nil {
		return err
	}
	fmt.Printf("task:      %s\n", d.Name)
	fmt.Printf("plugin:    %s\n", d.Plugin)
	fmt.Printf("status:    %s\n", statusText(&d.TaskSummary))
	fmt.Printf("interval:  %s\n", d.Interval)
	fmt.Printf("next run:  %s\n", orDash(d.NextRunAt))
	fmt.Printf("last sync: %s (success: %s)\n", orDash(d.LastSyncAt), orDash(d.LastSuccessAt))
	fmt.Printf("failures:  %d\n", d.ConsecutiveFailures)
	fmt.Printf("retention: keepLast=%d keepDays=%d\n", d.Retention.KeepLast, d.Retention.KeepDays)
	fmt.Printf("stats:     %d syncs, %d files, %s\n", d.Stats.TotalSyncs, d.Stats.TotalFiles, humanBytes(d.Stats.TotalBytes))
	return nil
}

func statusText(t *protocol.TaskSummary) string {
	if !t.Enabled {
		return "paused"
	}
	if t.Status == "syncing" {
		return "syncing"
	}
	if t.ConsecutiveFailures > 0 {
		return fmt.Sprintf("idle (%d fails)", t.ConsecutiveFailures)
	}
	return "idle"
}

func cmdSync(ctx ctx, name string) error {
	if name == "" {
		return fmt.Errorf("usage: edge-sync sync <task>")
	}
	fmt.Printf("syncing %s ...\n", name)
	var res struct {
		Triggered bool `json:"triggered"`
	}
	if err := ctx.ipc().call("task.trigger", map[string]any{"name": name}, &res); err != nil {
		return err
	}
	fmt.Printf("%s: sync completed\n", name)
	return nil
}

func cmdHistory(ctx ctx, name string) error {
	if name == "" {
		return fmt.Errorf("usage: edge-sync history <task>")
	}
	var versions []protocol.VersionInfo
	if err := ctx.ipc().call("history.list", map[string]any{"name": name}, &versions); err != nil {
		return err
	}
	fmt.Println(pad("VERSION", 20), pad("FILES", 8), pad("SIZE", 12), "NOTE")
	fmt.Println(hr(52))
	for _, v := range versions {
		note := ""
		if v.IsCurrent {
			note = "<- current"
		}
		fmt.Println(pad(v.Version, 20), pad(fmt.Sprint(v.Files), 8), pad(humanBytes(v.Bytes), 12), note)
	}
	fmt.Printf("\n%d version(s)\n", len(versions))
	return nil
}

func cmdFileVersions(ctx ctx, name, path string) error {
	if name == "" || path == "" {
		return fmt.Errorf("usage: edge-sync file-versions <task> <path>")
	}
	var versions []protocol.FileVersionEntry
	if err := ctx.ipc().call("history.fileVersions", map[string]any{"name": name, "path": path}, &versions); err != nil {
		return err
	}
	fmt.Println(pad("VERSION", 20), pad("SIZE", 12), "FINGERPRINT")
	fmt.Println(hr(52))
	for _, v := range versions {
		fp := v.Fingerprint
		if len(fp) > 14 {
			fp = fp[:14]
		}
		fmt.Println(pad(v.Version, 20), pad(humanBytes(v.Size), 12), fp)
	}
	return nil
}

func cmdLogs(ctx ctx, n int, task string) error {
	var entries []logLine
	if err := ctx.ipc().call("log.tail", map[string]any{"n": n, "task": task}, &entries); err != nil {
		return err
	}
	for _, e := range entries {
		fmt.Printf("%s %-5s %s %s\n", e.Time, strings.ToUpper(e.Level), e.Msg, e.Attrs)
	}
	return nil
}

type logLine struct {
	Time  string `json:"time"`
	Level string `json:"level"`
	Msg   string `json:"msg"`
	Attrs string `json:"attrs,omitempty"`
}

func cmdPlugins(ctx ctx) error {
	var plugins []protocol.PluginInfo
	if err := ctx.ipc().call("plugin.list", map[string]any{}, &plugins); err != nil {
		return err
	}
	for _, p := range plugins {
		fmt.Printf("%-12s v%s\n", p.Name, p.Version)
		var schema struct {
			Properties map[string]struct {
				Type        string `json:"type"`
				Description string `json:"description"`
			} `json:"properties"`
		}
		if len(p.ConfigSchema) > 0 && json.Unmarshal(p.ConfigSchema, &schema) == nil {
			for k, v := range schema.Properties {
				fmt.Printf("  - %s (%s)%s\n", k, orAny(v.Type), orEmpty(v.Description))
			}
		}
	}
	return nil
}

func cmdReload(ctx ctx) error {
	var res struct {
		Reloaded bool `json:"reloaded"`
		Tasks    int  `json:"tasks"`
	}
	if err := ctx.ipc().call("config.reload", map[string]any{}, &res); err != nil {
		return err
	}
	fmt.Printf("config reloaded (%d task(s))\n", res.Tasks)
	return nil
}

// ---------- add 向导 ----------

func cmdAdd(ctx ctx) error {
	var plugins []protocol.PluginInfo
	if err := ctx.ipc().call("plugin.list", map[string]any{}, &plugins); err != nil {
		return err
	}
	if len(plugins) == 0 {
		return fmt.Errorf("pluginBinDir 中未发现插件")
	}
	fmt.Println("available plugins:")
	for i, p := range plugins {
		fmt.Printf("  %d. %s (v%s)\n", i+1, p.Name, p.Version)
	}

	pick := ask("plugin (number or name): ")
	var chosen *protocol.PluginInfo
	for i, p := range plugins {
		if fmt.Sprint(i+1) == pick || p.Name == pick {
			chosen = &plugins[i]
			break
		}
	}
	if chosen == nil {
		return fmt.Errorf("invalid plugin choice")
	}

	name := ask("task name (letters/digits/._-): ")
	if !validTaskName(name) {
		return fmt.Errorf("invalid task name")
	}
	intervalSec := askDefault("poll interval (seconds): ", "600")
	keepLastStr := askDefault("retention keepLast (versions to keep): ", "10")
	keepDaysStr := askDefault("retention keepDays (0 = unlimited): ", "0")
	keepLast, keepErr := strconv.Atoi(keepLastStr)
	keepDays, daysErr := strconv.Atoi(keepDaysStr)
	if keepErr != nil || daysErr != nil || keepLast < 0 || keepDays < 0 {
		return fmt.Errorf("retention 值必须是非负整数（got keepLast=%q keepDays=%q）", keepLastStr, keepDaysStr)
	}

	options := map[string]any{}
	props := schemaProps(chosen.ConfigSchema)
	if len(props) > 0 {
		fmt.Printf("plugin %s options:\n", chosen.Name)
		for k, p := range props {
			raw := ask(fmt.Sprintf("  %s [%s]%s: ", k, p.Type, maybeDesc(p.Description)))
			if raw == "" {
				continue
			}
			options[k] = coerce(raw, p.Type)
		}
	} else {
		raw := ask("options (YAML/JSON, empty = {}): ")
		if raw != "" {
			if err := unmarshalLoose(raw, &options); err != nil {
				return err
			}
		}
	}

	task := map[string]any{
		"name":      name,
		"plugin":    chosen.Name,
		"interval":  intervalSec + "s",
		"options":   options,
		"retention": map[string]any{"keepLast": keepLast, "keepDays": keepDays},
	}
	fmt.Println()
	printYAMLTask(task)
	if !strings.EqualFold(ask("write this task to config? (y/N): "), "y") {
		fmt.Println("aborted")
		return nil
	}
	tasks := ctx.cfg.tasks()
	var kept []map[string]any
	for _, t := range tasks {
		if t["name"] != name {
			kept = append(kept, t)
		}
	}
	ctx.cfg.data["tasks"] = append(kept, task)
	if err := ctx.cfg.save(); err != nil {
		return err
	}
	if err := cmdReload(ctx); err != nil {
		return err
	}
	fmt.Printf("task \"%s\" added\n", name)
	return nil
}

func cmdEdit(ctx ctx, name string) error {
	if name == "" {
		return fmt.Errorf("usage: edge-sync edit <task>")
	}
	tasks := ctx.cfg.tasks()
	idx := -1
	for i, t := range tasks {
		if t["name"] == name {
			idx = i
			break
		}
	}
	if idx < 0 {
		return fmt.Errorf("task not found: %s", name)
	}
	editor := os.Getenv("EDITOR")
	if editor == "" {
		editor = "vi"
	}
	tmp := ctx.cfg.path + ".edit.yml"
	if err := os.WriteFile(tmp, []byte(fmt.Sprintf("# edit task %q; name must stay unchanged\n", name)+marshalTask(tasks[idx])), 0o644); err != nil {
		return err
	}
	cmd := exec.Command(editor, tmp)
	cmd.Stdin, cmd.Stdout, cmd.Stderr = os.Stdin, os.Stdout, os.Stderr
	if err := cmd.Run(); err != nil {
		os.Remove(tmp)
		return fmt.Errorf("editor exited nonzero, aborting")
	}
	edited, err := readYAML(tmp)
	os.Remove(tmp)
	if err != nil {
		return err
	}
	if edited["name"] != name {
		return fmt.Errorf("task name changed in editor, aborting")
	}
	tasks[idx] = edited
	ctx.cfg.data["tasks"] = tasks
	if err := ctx.cfg.save(); err != nil {
		return err
	}
	if err := cmdReload(ctx); err != nil {
		return err
	}
	fmt.Printf("task \"%s\" updated\n", name)
	return nil
}

func cmdRemove(ctx ctx, name string, purge bool) error {
	if name == "" {
		return fmt.Errorf("usage: edge-sync remove <task> [--purge]")
	}
	tasks := ctx.cfg.tasks()
	var kept []map[string]any
	found := false
	for _, t := range tasks {
		if t["name"] == name {
			found = true
			continue
		}
		kept = append(kept, t)
	}
	if !found {
		return fmt.Errorf("task not found: %s", name)
	}
	if purge {
		dir := filepath.Join(ctx.cfg.storagePath("dataDir"), name)
		fmt.Println("purging data directory:", dir)
		if err := os.RemoveAll(dir); err != nil {
			// 目录可能含 root 属主的历史残留；警告后继续移除任务配置。
			fmt.Printf("warning: purge incomplete (%v)；请用 root 手动清理残留\n", err)
		}
	}
	ctx.cfg.data["tasks"] = kept
	if err := ctx.cfg.save(); err != nil {
		return err
	}
	if err := cmdReload(ctx); err != nil {
		return err
	}
	fmt.Printf("task \"%s\" removed\n", name)
	return nil
}

// ---------- 取回 ----------

func cmdExport(ctx ctx, name, out string) error {
	if name == "" || out == "" {
		return fmt.Errorf("usage: edge-sync export <task> [--version <ts>] --out <dir>")
	}
	version := ctx.flags.Get("version")
	if version == "" {
		version = "current"
	}
	src := filepath.Join(ctx.cfg.storagePath("dataDir"), name)
	if version == "current" {
		src = filepath.Join(src, "current")
	} else {
		src = filepath.Join(src, "versions", version)
	}
	if _, err := os.Stat(src); err != nil {
		return err
	}
	if err := os.MkdirAll(out, 0o755); err != nil {
		return err
	}
	if err := copyTree(src, out); err != nil {
		return err
	}
	fmt.Printf("exported %s@%s -> %s\n", name, version, out)
	return nil
}

func cmdRestoreFile(ctx ctx, name, path, out string) error {
	if name == "" || path == "" || out == "" {
		return fmt.Errorf("usage: edge-sync restore-file <task> <path> [--version <ts>] --out <file>")
	}
	version := ctx.flags.Get("version")
	if version == "" {
		version = "current"
	}
	src := filepath.Join(ctx.cfg.storagePath("dataDir"), name)
	if version == "current" {
		src = filepath.Join(src, "current", filepath.FromSlash(path))
	} else {
		src = filepath.Join(src, "versions", version, filepath.FromSlash(path))
	}
	data, err := os.ReadFile(src)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(out), 0o755); err != nil {
		return err
	}
	if err := os.WriteFile(out, data, 0o644); err != nil {
		return err
	}
	fmt.Printf("restored %s@%s:%s -> %s (%d bytes)\n", name, version, path, out, len(data))
	return nil
}

// copyTree 递归复制目录（跳过版本元数据文件）。
func copyTree(src, dst string) error {
	entries, err := os.ReadDir(src)
	if err != nil {
		return err
	}
	for _, e := range entries {
		if e.Name() == manifestFileExport {
			continue
		}
		s := filepath.Join(src, e.Name())
		d := filepath.Join(dst, e.Name())
		if e.IsDir() {
			if err := os.MkdirAll(d, 0o755); err != nil {
				return err
			}
			if err := copyTree(s, d); err != nil {
				return err
			}
			continue
		}
		in, err := os.Open(s)
		if err != nil {
			return err
		}
		outF, err := os.Create(d)
		if err != nil {
			in.Close()
			return err
		}
		if _, err := io.Copy(outF, in); err != nil {
			in.Close()
			outF.Close()
			return err
		}
		in.Close()
		if err := outF.Close(); err != nil {
			return err
		}
	}
	return nil
}

const manifestFileExport = ".edge-sync-manifest.json"

func orDash(s string) string {
	if s == "" {
		return "-"
	}
	return s
}

func orAny(s string) string {
	if s == "" {
		return "any"
	}
	return s
}

func orEmpty(s string) string {
	if s == "" {
		return ""
	}
	return ": " + s
}

func validTaskName(name string) bool {
	if name == "" {
		return false
	}
	for i, r := range name {
		switch {
		case r >= 'a' && r <= 'z', r >= 'A' && r <= 'Z', r >= '0' && r <= '9':
		case r == '.' || r == '_' || r == '-':
			if i == 0 {
				return false
			}
		default:
			return false
		}
	}
	return true
}

func marshalTask(task map[string]any) string {
	raw, _ := json.MarshalIndent(task, "", "  ")
	return string(raw)
}

func printYAMLTask(task map[string]any) {
	fmt.Println(marshalTask(map[string]any{"tasks": []map[string]any{task}}))
}
